feature_tests/src/ownership.rs

```diff
@@ -0,0 +1,56 @@
+impl Drop for ffi::CountedOpaque {
+    fn drop(&mut self) {
+        self.0.fetch_sub(1, std::sync::atomic::Ordering::SeqCst);
+    }
+}
+
+#[diplomat::bridge]
+pub mod ffi {
+    use std::sync::atomic::{AtomicUsize, Ordering};
+    use std::sync::Arc;
+
+    /// Counts how many distinct `CountedOpaque` objects are instanciated
+    #[diplomat::opaque]
+    pub struct Counter(Arc<AtomicUsize>);
+
+    impl Counter {
+        pub fn new() -> Box<Counter> {
+            Box::new(Self(Arc::new(AtomicUsize::new(0))))
+        }
+
+        pub fn count(&self) -> usize {
+            self.0.load(Ordering::SeqCst)
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct CountedOpaque(pub(crate) Arc<AtomicUsize>);
+
+    impl CountedOpaque {
+        pub fn new(counter: &Counter) -> Box<CountedOpaque> {
+            counter.0.fetch_add(1, Ordering::SeqCst);
+            Box::new(Self(counter.0.clone()))
+        }
+    }
+
+    /// "Ownership is a delicous dish."
+    /// — OwnershipEater
+    #[diplomat::opaque]
+    pub struct OwnershipEater(Option<Box<CountedOpaque>>);
+
+    impl OwnershipEater {
+        pub fn new() -> Box<OwnershipEater> {
+            Box::new(Self(None))
+        }
+
+        // FIXME: not supported by C++ backend
```

Comment:
so the problem with taking ownership is that GC-based backends can't do this at all, I actually don't think we should be allowing this pattern in the first place.

As mentioned earlier, I feel like Diplomat non-opaque types should be `Copy`, no `Drop`, passed by-move, and typically constructed anew across FFI (with some potential caching as JS); and opaque types should be passed by-reference (only). Dealing with ownership otherwise, especially for GCd backends, is too much of a mess and often unsafe.

Currently this is a bit of a mess and not enforced, but I kinda want this to be the end state. What do you think?

Reply:
We can do manual ownership tracking as you're doing here but I'm not convinced we _should_. happy to be convinced otherwise

cc @sffc

Reply:
I think it’s fair to reduce the scope here, and moving is indeed not something most users of GC-based languages are expecting in the first place so it can be surprising.
I guess if user really need to move, the best way is to use an `Option` internally, on Rust side. Keeping the generated bindings less messy is good.